[JSC] eval() function from another realm shouldn't cause direct eval call

shvaikalesh opened the pull request:
#### 4737a82b02bd053262cbe1d0e2a503ec7ada627e
<pre>
[JSC] eval() function from another realm shouldn&apos;t cause direct eval call
<a href="https://bugs.webkit.org/show_bug.cgi?id=268027">https://bugs.webkit.org/show_bug.cgi?id=268027</a>
&lt;<a href="https://rdar.apple.com/problem/121546048">rdar://problem/121546048</a>&gt;

Reviewed by Yusuke Suzuki.

Whether eval() is direct or not is determined by SameValue() in [1] that compares the function
to `%eval%`, which denotes [2] specifically the eval() function of current global object,
not any built-in eval().

This change tightens the check for direct eval() and aligns JSC with V8 and SpiderMonkey.

Since Interpreter::eval() has tricky way of resolving global object, accounting for DFG inlining
code blocks, the callee check was moved there, ensuring all call sites handle empty JSValue.

[1]: <a href="https://tc39.es/ecma262/#sec-function-calls-runtime-semantics-evaluation">https://tc39.es/ecma262/#sec-function-calls-runtime-semantics-evaluation</a> (step 6.a)
[2]: <a href="https://tc39.es/ecma262/#sec-well-known-intrinsic-objects">https://tc39.es/ecma262/#sec-well-known-intrinsic-objects</a>

* JSTests/stress/direct-eval-cross-realm.js: Added.
* JSTests/test262/expectations.yaml: Mark 1 test as passing.
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::commonCallDirectEval):
* Source/JavaScriptCore/runtime/JSFunctionInlines.h:
(JSC::isHostFunction): Deleted.

Canonical link: <a href="https://commits.webkit.org/273782@main">https://commits.webkit.org/273782@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c12faeadd1eafd070157840f17d01e5317d2a4a6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/36657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/15599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/38886 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/18072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/12676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/39314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/37218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/18072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/38886 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/18072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/38886 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/40561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/30766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/18072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/38886 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/40561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/36429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/11787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/12676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/40561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/38886 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/43216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/43216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4742 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/12674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -2045,7 +2045,7 @@ JSC_DEFINE_JIT_OPERATION(operationCallDirectEvalStrict, EncodedJSValue, (void* f
     calleeFrame->setCodeBlock(nullptr);
 
-    if (!isHostFunction(calleeFrame->guaranteedJSValueCallee(), globalFuncEval))
+    if (calleeFrame->guaranteedJSValueCallee() != calleeFrame->callerFrame()->codeBlock()->globalObject()->evalFunction())
```
Is it correct even if the Eval is inside inlined CodeBlock from different realm's DFG CodeBlock?

shvaikalesh replied:
Oh, `Interpreter::eval()` resolved global object in way more complicated way than `calleeFrame->callerFrame()->codeBlock()->globalObject()`!!

I moved the check there.

Constellation reviewed (commented):
Commented

Constellation reviewed (approved):
r=me

Commit:
[JSC] eval() function from another realm shouldn't cause direct eval call
https://bugs.webkit.org/show_bug.cgi?id=268027
<rdar://problem/121546048>

Reviewed by Yusuke Suzuki.

Whether eval() is direct or not is determined by SameValue() in [1] that compares the function
to `%eval%`, which denotes [2] specifically the eval() function of current global object,
not any built-in eval().

This change tightens the check for direct eval() and aligns JSC with V8 and SpiderMonkey.

Since Interpreter::eval() has tricky way of resolving global object, accounting for DFG inlining
code blocks, the callee check was moved there, ensuring all call sites handle empty JSValue.

[1]: https://tc39.es/ecma262/#sec-function-calls-runtime-semantics-evaluation (step 6.a)
[2]: https://tc39.es/ecma262/#sec-well-known-intrinsic-objects

* JSTests/stress/direct-eval-cross-realm.js: Added.
* JSTests/test262/expectations.yaml: Mark 1 test as passing.
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::commonCallDirectEval):
* Source/JavaScriptCore/runtime/JSFunctionInlines.h:
(JSC::isHostFunction): Deleted.

Canonical link: https://commits.webkit.org/273782@main
